primitives/runtime/src/lib.rs

```diff
@@ -95,7 +95,52 @@ pub use either::Either;
 /// the block itself would allow swapping justifications to change the block's hash
 /// (and thus fork the chain). Sending a `Justification` alongside a block instead
 /// bypasses this problem.
-pub type Justification = Vec<u8>;
+pub type EncodedJustification = Vec<u8>;
+
+/// Justification together with an ID to tag it's origin.
+pub type Justification = (ConsensusEngineId, EncodedJustification);
+
+/// Collection of Justifications, since we might have more than one stored per block.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct Justifications(Vec<Justification>);
```

Comment:
Would it be possible to use a `HashSet` instead of a `Vec` here? This would be an extra saveguard against duplicate justifications in addition to the `append_justification` method checking for this condition.

Reply:
We want to check against duplicate keys, not items, so a map structure might be a better fit. That said, if I recall correctly the `sp_std::Vec` was chosen for simplicity and avoiding a `std` dependency in primitives. If this is crucial for something that is (currently) only used by the client I'm not entirely sure though.
This vec should in practice never have more than 1 or 2 items, so a Vec seemed simple, but yeah I understand the alure of map structure :)